BPA/Ch.06/M0.stan

```diff
@@ -1,45 +1,46 @@
 data {
-  int<lower=0> M;               // Size of augumented data set
-  int<lower=0> T;               // Number of sampling occasions
-  int<lower=0,upper=1> y[M, T]; // Capture-history matrix
+  int<lower=0> M; // Size of augumented data set
+  int<lower=0> T; // Number of sampling occasions
+  array[M, T] int<lower=0, upper=1> y; // Capture-history matrix
 }
-
 transformed data {
-  int<lower=0> s[M];            // Totals in each row
-  int<lower=0> C;               // Size of observed data set
-
+  array[M] int<lower=0> s; // Totals in each row
+  int<lower=0> C; // Size of observed data set
+  
   C = 0;
-  for (i in 1:M) {
+  for (i in 1 : M) {
     s[i] = sum(y[i]);
-    if (s[i] > 0)
+    if (s[i] > 0) 
       C = C + 1;
   }
 }
-
 parameters {
-  real<lower=0,upper=1> omega;  // Inclusion probability
-  real<lower=0,upper=1> p;      // Detection probability
+  real<lower=0, upper=1> omega; // Inclusion probability
+  real<lower=0, upper=1> p; // Detection probability
 }
-
 model {
   // Priors are imlicitly defined;
   //  omega ~ uniform(0, 1);
   //  p ~ uniform(0, 1);
-
+  
   // Likelihood
-  for (i in 1:M)
-    if (s[i] > 0)
-      // z[i] == 1
-      target += bernoulli_lpmf(1 | omega)
-              + binomial_lpmf(s[i] | T, p);
-    else // s[i] == 0
-      target += log_sum_exp( bernoulli_lpmf(1 | omega)   // z[i] == 1
-                             + binomial_lpmf(0 | T, p),
-                             bernoulli_lpmf(0 | omega)); // z[i] == 0
+  for (i in 1 : M) 
+    if (s[i] > 0) 
+    // z[i] == 1
+      target += bernoulli_lpmf(1 | omega) + binomial_lpmf(s[i] | T, p);
+    // s[i] == 0
+    
+    else 
+    
+      target += log_sum_exp(bernoulli_lpmf(1 | omega)
+                            + binomial_lpmf(0 | T, p),
+                            bernoulli_lpmf(0 | omega)); // z[i] == 0
+  
+  // ^^^: z[i] == 1
```

Comment:
@nhuurre - related to https://github.com/stan-dev/stanc3/pull/894, do you think there is any way we could handle this better when auto-formatting?

Reply:
The bad formatting here is 
1. too much space around `else` 
2. `s[i] == 0` comment on the wrong side of `else`
3.  `z[i] == 1` comment pushed out completely

(1) is probably not too bad, just requires some tweaking `pp_recursive_ifthenelse`. (2) is annoying because the AST does not record keyword locations. Could either change the AST or record keywords in the comment list (like commas). (3) is a case of comments inside binary operator expressions being skipped. Probably easy to fix, the printer code should be very similar to the code for expression lists in `pp_list_of`. (but again, needs a hack to find the exact location of the operator token)

Btw, manually adding braces and moving the `z[i]` comment to the next line makes the auto-format produce
```stan
  for (i in 1 : M) 
    if (s[i] > 0) {
      // z[i] == 1
      target += bernoulli_lpmf(1| omega) + binomial_lpmf(s[i]| T, p);
    } else {
      // s[i] == 0
      target += log_sum_exp(bernoulli_lpmf(1| omega) + binomial_lpmf(0| T, p)
                            // z[i] == 1
                            , bernoulli_lpmf(0| omega)); // z[i] == 0
    }
```

Reply:
I think it is worthwhile to make edits to the AST to retain more location information as needed. It's probably not bad to have around anyway and aligns with the "AST is intended to have a direct 1-1 mapping with the syntax, so there are things like parentheses being kept around" goal of stanc3, IMO. I also think it's better to have in the AST than in the comment list like we're currently doing for commas, but I can be persuaded either way.

I agree that your formatting with the braces is preferrable, but I think this is also a decent example of the kind of thing someone has written 'in the wild', and it would be good to support as much as possible. Would you be able to take a look at this some more? Even just fixing the "easy" parts you outlined would be a great plus